src/openstack_cloud/openstack_manager.py

```diff
@@ -0,0 +1,378 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for handling interactions with OpenStack."""
+import json
+import logging
+import secrets
+from dataclasses import dataclass
+from typing import Iterable, Literal, Optional
+
+import jinja2
+import keystoneauth1.exceptions.http
+import openstack
+import openstack.compute.v2.server
+import openstack.connection
+import openstack.exceptions
+import openstack.image.v2.image
+from openstack.exceptions import OpenStackCloudException
+from openstack.identity.v3.project import Project
+
+from charm_state import Arch, ProxyConfig, SSHDebugConnection, UnsupportedArchitectureError
+from errors import (
+    OpenstackImageBuildError,
+    OpenStackUnauthorizedError,
+    RunnerBinaryError,
+    SubprocessError,
+)
+from github_client import GithubClient
+from github_type import RunnerApplication
+from runner_type import GithubPath
+from utilities import execute_command, retry
+
+logger = logging.getLogger(__name__)
+
+IMAGE_PATH_TMPL = "jammy-server-cloudimg-{architecture}-compressed.img"
+IMAGE_NAME = "jammy"
+BUILD_OPENSTACK_IMAGE_SCRIPT_FILENAME = "scripts/build-openstack-image.sh"
+
+
+def _create_connection(cloud_config: dict[str, dict]) -> openstack.connection.Connection:
```

Comment:
We might need to manage the openstack connection differently.

The `openstack.connection.Connection` has defined the `__enter__`, and `__exit__` method, therefore it is likely designed to be used a context manager. 
The `__exit__` includes a call to `close` which does some cleanup: https://github.com/openstack/openstacksdk/blob/601b21acadf30f54e158523b48af47d3cac23fa5/openstack/connection.py#L654
We should be ensure `close`\`__exit__` is called at the end of each connection.

Also we would need to convert all `keystoneauth1.exceptions.http.Unauthorized` to `OpenStackUnauthorizedError` so the `charm.py` can just deal with `OpenStackUnauthorizedError` and not leak implementation detail to `charm.py`.

I was planning to do something like this:
```
from contextlib import contextmanager

@contextmanager
def _create_connection(self) -> Iterator[OpenstackConnection]:
    try:
        with OpenstackConnection(self.cloud_name) as conn:
            yield conn
    except keystoneauth1.exceptions.http.Unauthorized as exc:
        logger.exception("Unable to connect to openstack")
        raise OpenStackUnauthorizedError() from exc
```

Which would be used like:
```
with self._create_connection as conn:
    # conn is a openstack.connection.Connection instance.
    # Use conn to call openstack API.
```

Reply:
Refactored, thank you!
I don't think instantiating `OpenstackConnection` would raise - hence I've added `conn.authorize()` before yielding the connection so that we can catch the `openstack.exceptions.HttpException` (notice that it isn't keystoneauth exception) before we let the users use the connection.